packages/react-router/src/Matches.tsx

```diff
@@ -27,8 +27,75 @@ import type {
 } from './routeInfo'
 import type { ControlledPromise, DeepPartial, NoInfer } from './utils'
 
+export type MatchSuggestions<
+  TValue,
+  TParentPath extends string,
+> = TValue extends string | number | bigint | boolean | symbol
+  ? never
+  : TValue extends ReadonlyArray<any>
+    ? `${TParentPath}${keyof TValue & `${number}`}`
+    : `${TParentPath}${keyof TValue & string}`
+
+export type FindMatchByKey<TMatch, TKey, TValue> = TValue extends any
+  ? TKey extends keyof TValue
+    ? TMatch
+    : never
+  : never
+
+export type ValueOf<TKey, TValue> = TValue extends any
+  ? TKey extends keyof TValue
+    ? TValue[TKey]
+    : never
+  : never
+
+export type SuggestMatchByPath<
+  TPath,
+  TMatch,
+  TValue = TMatch,
+  TParentPath extends string = '',
+> = TPath extends `${string}.${string}`
+  ? TPath extends `${infer TFirst}.${infer TRest}`
+    ? SuggestMatchByPath<
+        TRest,
+        TMatch,
+        ValueOf<TFirst, TValue>,
+        `${TParentPath}${TFirst}.`
+      >
+    : never
+  : {
+      suggestions: MatchSuggestions<TValue, TParentPath>
+      match: FindMatchByKey<TMatch, TPath, TValue>
+    }
+
+export type SuggestMatchesByPath<TPath, TMatch> = TMatch extends any
+  ? SuggestMatchByPath<TPath, TMatch>
+  : never
+
+export type ValidateMatchSuggestions<
+  TMatch extends AnyRouteMatch,
+  TPath extends string,
+> = TPath extends SuggestMatchesByPath<TPath, TMatch>['suggestions']
+  ? TPath
+  : SuggestMatchesByPath<TPath, TMatch>['suggestions']
+
+export const isMatch = <TMatch extends AnyRouteMatch, TPath extends string>(
+  match: TMatch,
+  path: ValidateMatchSuggestions<TMatch, TPath>,
+): match is SuggestMatchesByPath<TPath, TMatch>['match'] => {
+  const parts = path.split('.')
```

Comment:
This runtime logic is really basic. I'm going to be adding tests so I assume it might get better